pybamm/solvers/idaklu_jax.py

```diff
@@ -0,0 +1,956 @@
+
+
+class IDAKLUJax:
+    """
+    JAX wrapper for IDAKLU solver
+
+    Objects of this class should be created via an IDAKLUSolver object.
+
+    Parameters
+    ----------
+    solver : :class:`pybamm.IDAKLUSolver`
+        The IDAKLU solver object to be wrapped
+    """
+
+    def __init__(
+        self,
+        solver,
+    ):
+        if not pybamm.have_jax():
+            raise ModuleNotFoundError(
+                "Jax or jaxlib is not installed, please see https://docs.pybamm.org/en/latest/source/user_guide/installation/gnu-linux-mac.html#optional-jaxsolver"
+            )  # pragma: no cover
+        if not pybamm.have_idaklu():
+            raise ModuleNotFoundError(
+                "IDAKLU is not installed, please see https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html"
+            )  # pragma: no cover
+        self.jaxpr = None  # JAX expression
+        self.idaklu_jax_obj = None  # IDAKLU-JAX object
+        self.solver = solver  # Originating IDAKLU Solver object
+
+    def get_jaxpr(self):
+        """Returns a JAX expression representing the IDAKLU-wrapped solver object"""
+        if self.jaxpr is None:
+            raise pybamm.SolverError("jaxify() must be called before get_jaxpr()")
+        return self.jaxpr
+
+    def get_var(
+        self,
+        f: Callable,
+        varname: str,
+    ):
+        """Helper function to extract a single variable from the jaxified expression
+
+        Returns a JAX expression
+
+        Parameters
+        ----------
+        f : function
+            The jaxified expression
+        varname : str
+            The name of the variable to extract
+        """
+
+        def f_isolated(*args, **kwargs):
+            out = f(*args, **kwargs)
+            index = self.jax_output_variables.index(varname)
+            if out.ndim == 0:
+                return out  # pragma: no cover
+            elif out.ndim == 1:
+                return out[index]
+            else:
+                return out[:, index]
+
+        return f_isolated
+
+    def get_vars(
+        self,
+        f: Callable,
+        varnames: List[str],
+    ):
+        """Helper function to extract multiple variables from the jaxified expression
+
+        Returns a JAX expression
+
+        Parameters
+        ----------
+        f : function
+            The jaxified expression
+        varnames : list of str
+            The names of the variables to extract
+        """
+
+        def f_isolated(*args, **kwargs):
+            out = f(*args, **kwargs)
+            index = np.array(
+                [self.jax_output_variables.index(varname) for varname in varnames]
+            )
+            if out.ndim == 0:
+                return out  # pragma: no cover
+            elif out.ndim == 1:
+                return out[index]
+            else:
+                return out[:, index]
+
+        return f_isolated
+
+    def jax_value(
+        self,
+        *,
```

Comment:
I've not seen this, what does the `*` do here?

Reply:
A bare asterisks prevents any subsequent arguments from being callled positionally; this forces them to be named and so is useful to prevent misordering based on their position alone (see also https://stackoverflow.com/a/14302007).

Reply:
nice :)